### PR TITLE
Let folding

### LINKED ISSUE #1899: Let-folding
Implement an optimizing transformation that folds lets whose values are immediate, i.e., they don't require evaluation (variables or constants).

For example, transform
```
let x := y in let z := x + x in x + z
```
to
```
let z := y + y in y + z
```

This is easy to implement using the `rmap` recursor and will remove the spurious lets created by the pattern matching compiler (both the naive and the new proper one).

### PR DESCRIPTION
**Pull request by lukaszcz**
* Closes #1899

### CODE REVIEW

File: src/Juvix/Compiler/Core/Extra/Utils.hs [hunk 1]
```diff
@@ -39,6 +39,13 @@ isTypeConstr tab ty = case typeTarget ty of
     isTypeConstr tab (fromJust $ HashMap.lookup _identSymbol (tab ^. identContext))
   _ -> False
 
+isImmediate :: Node -> Bool
```
Comment (janmasrovira) [inline src/Juvix/Compiler/Core/Extra/Utils.hs:42]: add a short description

File: src/Juvix/Compiler/Core/Transformation/Optimize/LetFolding.hs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+module Juvix.Compiler.Core.Transformation.Optimize.LetFolding where
```
Comment (janmasrovira) [inline src/Juvix/Compiler/Core/Transformation/Optimize/LetFolding.hs:1]: add a short description of the transformation at the top of the module

File: src/Juvix/Compiler/Core/Extra/Utils.hs (hunk 1 above)
Comment (lukaszcz) [reply src/Juvix/Compiler/Core/Extra/Utils.hs:44]: done

File: src/Juvix/Compiler/Core/Transformation/Optimize/LetFolding.hs (hunk 2 above)
Comment (lukaszcz) [reply src/Juvix/Compiler/Core/Transformation/Optimize/LetFolding.hs:13]: done

**Review by paulcadman — APPROVED** [review]

**Pull request merged**